six/modules/c++/cphd/include/cphd/Data.h

```diff
@@ -20,86 +20,314 @@
+        return channels.size();
+    }
+    size_t getNumBytesPVPSet() const
+    {
+        return numBytesPVP;
+    }
+    std::string getCompressionID() const
+    {
+        return signalCompressionID;
+    }
+    SignalArrayFormat getSignalFormat() const
+    {
+        return signalArrayFormat;
+    }
+    size_t getNumSupportArrays() const
+    {
+        return supportArrayMap.size();
+    }
+    //! Get specific support array by identifier
+    SupportArray getSupportArrayById(const std::string& id) const;
 
-    bool operator!=(const Data& other) const
+    //! Get size of support array element
+    size_t getElementSize(const std::string& id) const
     {
-        return !((*this) == other);
+        return getSupportArrayById(id).bytesPerElement;
     }
+    size_t getAllSupportSize() const;
+
+    //! Create and add new support array
+    //! Validates, and add new support array to supportArrayMap
+    void setSupportArray(const std::string& id, size_t numRows,
+                         size_t numCols, size_t numBytes,
+                         sys::Off_T offset);
+
+    //! Checks if wideband data is compressed
+    bool isCompressed() const;
 
-    SampleType sampleType;
-    size_t numCPHDChannels;
-    size_t numBytesVBP;
+public:
+    //! Signal Array sample binary format of the CPHD
+    //! signal arrays in standard (i.e. uncompressed)
+    //! format.
+    SignalArrayFormat signalArrayFormat;
+
+    //! Number of bytes per set of Per Vector
+    //! Parameters. One set of PVPs for each CPHD
+    //! signal vector.
+    size_t numBytesPVP;
+
+    //! Parameters that define the Channel signal array
+    //! and PVP array size and location
+    std::vector<Channel> channels;
+
+    //! (Optional) Parameter that indicates the signal arrays are in
+    //! compressed format. Value identifies the method
+    //! of decompression
+    std::string signalCompressionID;
+
+    // Both of these maps get populated and/or edited together
+    // Made for O(logn) lookup
+    //! (Optional) Map of unique support array id to support array offset
+    // std::unordered_map<std::string,sys::Off_T> supportOffsetMap;
+    // (Optional) Ordered map to store support arrays with offset key
+    // std::map<sys::Off_T,SupportArray, CmpByOffset> supportArrayMap;
+
+    // (Optional) Unordered map key: unique identifier string,
+    // value: supportArray object
+    std::unordered_map<std::string, SupportArray> supportArrayMap;
+
+private:
+    // Custom comparator
+    struct CmpByOffset
+    {
+        // Overload operator ()
+        // Returns true if offset lhs is less than other offset
+        bool operator()(const sys::Off_T& lhs, const sys::Off_T& rhs) const
+        {
+            return lhs < rhs;
+        }
+    };
 
-    std::vector<ArraySize> arraySize;
+    // Book keeping map for efficient validation
+    // Support Array Map with:
+    // key: offset, value: array size
+    std::map<sys::Off_T, size_t, CmpByOffset> mOffsetMap;
```

Comment:
This ordered map is private and is only used for O(logn) time validation when adding to the supportArrayMap.
I can remove it in favor of simplicity but validation time complexity will become O(n)

Reply:
This looks a lot better! `std::map` sorts the entries in ascending order by default, so you can get rid of the custom comparator as well.